### PR TITLE
TorchDynamo: Add convolution binary(inplace) fusion for cpu in inference mode

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88864
* #88794
* #88414
* #88412
* __->__ #88403

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/overrides.py [hunk 1]
```diff
@@ -473,6 +561,56 @@ def fuse_binary(gm: torch.fx.GraphModule):
     return gm
 
 
+def fuse_binary_inplace(gm: torch.fx.GraphModule):
```
Comment (jgong5) [inline torch/_inductor/overrides.py:564]: Can we share code with `fuse_binary`?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -3447,6 +3447,71 @@ def apply_constraint(self):
         self.freeze_layout_with_stride_order(self.layout.preferred_stride_order)
 
 
+class ConvolutionBinaryInplace(ExternKernelAlloc):
```
Comment (jgong5) [inline torch/_inductor/ir.py:3450]: Can we share code with `ConvolutionBinary`?

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -3447,6 +3447,71 @@ def apply_constraint(self):
...
+            f"{self.get_name()} = {self.kernel}({', '.join(self.codegen_args())})"
+        )
+        if isinstance(self.layout, Layout):
+            self.codegen_size_asserts(wrapper)
+
+    @classmethod
+    def create(
+        cls,
+        x: "TensorBox",
+        other: "TensorBox",
```
Comment (jgong5) [inline torch/_inductor/ir.py:3474]: should take care of the in-place MutationLayout.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/overrides.py (hunk 1 above)
Comment (XiaobingSuper) [reply torch/_inductor/overrides.py:571]: For inplace case, we have some special checks, I simplified those codes by abstracting some common code.

File: torch/_inductor/ir.py (hunk 2 above)
Comment (XiaobingSuper) [reply torch/_inductor/ir.py:3470]: in-place op is different from outplace cases, I think it is ok to Individually define it.

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -1690,7 +1690,7 @@ def __init__(self, target: IRNode):
             target.get_device(),
             target.get_dtype(),
             target.get_size(),
-            None,  # type: ignore[arg-type]
+            target.get_stride(),
```
Comment (XiaobingSuper) [inline torch/_inductor/ir.py:1693]: Is it make sense to add stride type? now, if we don't add such a parameter, such as a inplace op's post op is convolution, we will call **require_stride1** which will get input's stride and will assert an error.

Comment (jansel) [reply torch/_inductor/ir.py:1693]: I think there is an issue where the thing we are mutating might be a FlexibleLayout, which means the stride can change after we call this constructor, so copying strides here load lead to subtle bugs..  We could add a `@property` that redirects to the true layout.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -1694,6 +1706,16 @@ def __init__(self, target: IRNode):
         )
         self.target = target
 
+    @Layout.stride.getter
+    def stride(self):
+        return self.real_layout.stride
+
+    @property
+    def real_layout(self):
```
Comment (XiaobingSuper) [inline torch/_inductor/ir.py:1714]: @jansel , for MutationLayout's stride, I will query target's layout to get the real layout, is it ok?

Comment (jansel) [reply torch/_inductor/ir.py:1751]: Yes this seems fine, let's make this a method not a property though.

Comment (XiaobingSuper) [reply torch/_inductor/ir.py:1750]: Changed.

**Comment by XiaobingSuper** [comment]
@jansel @Chillee, this PR use mutation layout for an in-place external op,  hoping it doesn't introduce an issue.

File: torch/_inductor/ir.py (hunk 5 above)
Comment (jgong5) [reply torch/_inductor/ir.py:1786]: Maybe naming it `target_layout` would be more consistent?

File: torch/_inductor/overrides.py [hunk 6]
```diff
@@ -473,6 +568,48 @@ def fuse_binary(gm: torch.fx.GraphModule):
...
+    modules = dict(gm.named_modules())
+    for node in gm.graph.nodes:
+        if check_node_is_add_inplace(node) and (
+            len(node.kwargs) != 2 or node.kwargs["alpha"] == 1.0
```
Comment (jgong5) [inline torch/_inductor/overrides.py:575]: We need a better way of checking whether the `add` op has `out` kwarg or not. The check with `len(node.kwargs)` seems not straightforward. Can we check the existence of `out` in kwargs directly?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/overrides.py (hunk 6 above)
Comment (XiaobingSuper) [reply torch/_inductor/overrides.py]: Yes, I updated the check code,

**Review by jgong5 — APPROVED** [review]

File: torch/_inductor/ir.py [hunk 7]
```diff
@@ -2395,6 +2417,15 @@ def require_stride_order(cls, x, order):
             order
         ):
             return x
+        elif isinstance(x.get_layout(), MutationLayout):
+            if isinstance(x.layout.real_layout, FlexibleLayout):
+                # if MutationLayout's real layout is FlexibleLayout, do we need fix it as a FixedLayout?
+                # as_storage_and_layout(x, freeze=True, want_contiguous=False, stride_order=order)
+                return x
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_inductor/ir.py]: Yes

File: torch/_inductor/ir.py [hunk 8]
```diff
@@ -2462,6 +2483,15 @@ def require_stride_order(cls, x, order):
                 x.get_layout(), FixedLayout
             ) and x.get_layout().is_stride_ordered(order):
                 return x
+            elif isinstance(x.get_layout(), MutationLayout):
+                if isinstance(x.get_layout().real_layout(), FlexibleLayout):
+                    # if MutationLayout's real layout is FlexibleLayout, do we need fix it as a FixedLayout?
+                    # as_storage_and_layout(x, freeze=True, want_contiguous=False, stride_order=order)
+                    return x
```
Comment (jansel) [inline torch/_inductor/ir.py:2490]: This seems wrong, though I am not sure we will hit this case.  Perhaps add an assert we don't hit this case?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/ir.py (hunk 8 above)
Comment (XiaobingSuper) [reply torch/_inductor/ir.py]: I added an assert message here which will not hit this case.

**Review by jansel — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**